Java/core/src/main/java/com/amazon/randomcutforest/store/IntervalManager.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.store;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.checkState;
+
+import java.util.Arrays;
+
+/**
+ * This class defines common functionality for Store classes, including
+ * maintaining the stack of free pointers.
+ */
+public class IntervalManager {
+
+    protected int capacity;
+    protected int[] freeIndexesStart;
+    protected int[] freeIndexesEnd;
+    protected int lastInUse;
+
+    public IntervalManager(int capacity) {
+        checkArgument(capacity > 0, "incorrect parameters");
+        freeIndexesEnd = new int[1];
+        freeIndexesStart = new int[1];
+        lastInUse = 1;
+        this.capacity = capacity;
+        freeIndexesStart[0] = 0;
+        freeIndexesEnd[0] = capacity - 1;
+    }
+
+    public void extendCapacity(int newCapacity) {
+        checkArgument(newCapacity > capacity, " incorrect call, we can only increase capacity");
```

Comment:
Is this check correct? Looks like it should be `<`.

Reply:
for the index managers, the capacity is not the maximum capacity but a current capacity which we kept increasing in RCF2.0. We may be able to eliminate that need with 3.0 altogether (because of the intervals).